### PR TITLE
Implement last baseline alignment for Flexbox containers.

### PR DESCRIPTION
**Pull request by sammygill**
#### a7e014a014305d26b5504f91e940b56e94848714
<pre>
Implement last baseline alignment for Flexbox containers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244734">https://bugs.webkit.org/show_bug.cgi?id=244734</a>
rdar://99506835

Reviewed by Brent Fulgham.

In order to implement last baseline alignment for flex containers, we
need to hold a little bit more information. Currently,
RenderFlexibleBox::LineContext holds a maxAscent value, which holds the
value of the furthest baseline in the cross axis. This is then used to
shift the other flex items to this baseline. A new field called
lastBaselineMaxAscent was added to achieve this same functionality for
the last baseline elements. This needs to be a separate field in
scenarios where you may have certain items being aligned by their
baselines and others by their last baselines.

This patch also adds the “margin before” to the first baseline of table
elements since it was not being added before. This is needed in order to
properly align other elements alongside a table with margins. Without
this addition, we were rendering the expectation for both
css-flexbox/flexbox-align-self-baseline-horiz-001a and
css-flexbox/flexbox-align-self-baseline-horiz-001b incorrectly.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::LineContext::LineContext):
(WebCore::RenderFlexibleBox::marginBoxAscentForChild):
Since marginBoxAscentForChild is being called during alignment purposes,
we can determine what type of alignment is being done here and return
the appropriate ascent (first baseline vs last baseline).

(WebCore::alignmentOffset):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
Depending on whether the item is being first baseline aligned or last
baseline aligned, we need to either update maxAscent or
lastBaselineMaxAscent. Later on when the alignment is being done, we
can check the type of alignment one more time and use the correct
value.

(WebCore::RenderFlexibleBox::alignChildren):
Since items need to be pushed to the end of the cross axis when
wrap-reverse is specified, we can use the same logic to push the items
that are being last baseline aligned. There was also a little bit of
cleanup that I was able to do here. Instead of waiting until all of the
lines have had their items aligned to make any adjustments (e.g. due
to wrap-reverse), we can do it after laying out each line. We can
determine whether or not we need to loop over the line again by checking
the value of minMarginAfterBaseline. If this value was set then that
means that not only is there room to move the items in the flex line,
but also that we need to because the flex container has wrap-reverse
set or some items are being last baseline aligned.

(WebCore::RenderFlexibleBox::childIsParticipatingInBaselineAlignment const):
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTable::baselinePosition const):
This is not directly related to the last baseline implementation but to
a test case for it. There were 2 expectation files we were rendering
incorrectly
(css-flexbox/flexbox-align-self-baseline-horiz-001a-expected.xhtml and
flexbox-align-self-baseline-horiz-001b-expected.xhtml).

These were using the first baseline for table elements and we were not
computing the correct first baseline. This caused us to fail the tests
even though we were rendering the actual test file correctly.

Canonical link: <a href="https://commits.webkit.org/255383@main">https://commits.webkit.org/255383@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c1e411a265067e248fd4ee6ab7b65eedd64c6405

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102098 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1550 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29938 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84752 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98261 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1021 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78839 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27955 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82613 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71007 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36356 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34119 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17734 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3731 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40360 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39893 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36876 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderBlockFlow.cpp [hunk 1]
```diff
@@ -2998,6 +2998,25 @@ std::optional<LayoutUnit> RenderBlockFlow::firstLineBaseline() const
...
+
+    if (!childrenInline())
+        return RenderBlock::lastLineBaseline();
+
+    if (!hasLines())
+        return std::nullopt;
+
+    const LayoutIntegration::LineLayout* lineLayout = modernLineLayout();
+    ASSERT(lineLayout); 
+    return LayoutUnit { floorToInt(lineLayout->lastLineLogicalBaseline()) };
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderBlockFlow.cpp]: Why does this use lastLineLogicalBaseline and  RenderBlockFlow::firstLineBaseline uses firstLinePhysicalBaseline? What about the last few lines of RenderBlockFlow::firstLineBaseline, why is that not needed here?

File: Source/WebCore/rendering/RenderBlock.cpp [hunk 2]
```diff
@@ -2555,6 +2555,23 @@ std::optional<LayoutUnit> RenderBlock::firstLineBaseline() const
     return std::optional<LayoutUnit>();
 }
 
+std::optional<LayoutUnit> RenderBlock::lastLineBaseline() const
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderBlock.cpp]: This is very similar to RenderBlock::firstLineBaseline(), could it be shared, maybe by passing a lambda?

File: Source/WebCore/rendering/RenderBlockFlow.cpp (hunk 1 above)
Comment (alanbaradlay) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: We need a LineLayout::lastLinePhysicalBaseline() here and it should also be moved behind the LineBox iterator interface.
1. you can't assume all the content goes through the modern line layout codepath
2. It would also eliminate the need for duplicating the vertical flip (what Rob refers to) since that would be moved behind the iterator as well.
Not sure why we haven't done that yet.

Comment (sammygill) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: I implemented LineLayout::lastLinePhysicalBaseline and added some extra logic for the case where the code goes through legacy layout. After talking with Alan, it seems like this should be ok for now, but eventually we will want to do this refactoring. Things will also probably change as we move more away from legacy line layout.

File: Source/WebCore/rendering/RenderBlock.cpp (hunk 2 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBlock.cpp]: I'm not sure if a lambda would have been easy to do since that would probably need to be done at the base class and not all derived implementations may need it. What I ended up doing was implementing RenderBox::firstInFlowChildBox and RenderBox::lastInFlowChildBox to help clean this up. I think it the overall readability and reusability, but let me know what you think!

**Comment by sammygill** [comment]
Updated RenderBlockFlow::lastLineBaseline so that the call to modern layout lastLinePhysicalBaseline was guarded by an if instead of an assert. Now we should fall back to legacy line layout the scenarios where modernLineLayout returns a nullptr

File: Source/WebCore/rendering/RenderBlock.cpp [hunk 3]
```diff
@@ -2546,13 +2546,25 @@ std::optional<LayoutUnit> RenderBlock::firstLineBaseline() const
...
+{
+    if (shouldApplyLayoutContainment())
+        return std::nullopt;
+
+    if (isWritingModeRoot() && !isRubyRun())
+        return std::optional<LayoutUnit>();
+
+    for (RenderBox* child = lastInFlowChildBox(); child; child = child->nextInFlowSiblingBox()) {
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderBlock.cpp]: This loop doesn't make much sense, if `child` is initialized to the last box, then `child->nextInFlowSiblingBox()` should be null?

File: Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h [hunk 4]
```diff
@@ -33,6 +33,7 @@
 #include "LayoutPoint.h"
 #include "LayoutState.h"
 #include "RenderObjectEnums.h"
+#include "layout/integration/inline/LayoutIntegrationLine.h"
```
Comment (Loirooriol) [inline Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h]: Would it be possible to use a forward declaration instead?

File: Source/WebCore/rendering/RenderTableSection.cpp [hunk 5]
```diff
@@ -874,9 +874,13 @@ std::optional<LayoutUnit> RenderTableSection::firstLineBaseline() const
...
         return std::optional<LayoutUnit>();
 
+    LayoutUnit tableMarginBefore;
+    if (RenderTable* parentTable = table())
+        tableMarginBefore = parentTable->style().marginBefore().value();
+
     LayoutUnit firstLineBaseline = m_grid[0].baseline;
     if (firstLineBaseline)
-        return firstLineBaseline + m_rowPos[0];
+        return firstLineBaseline + m_rowPos[0] + tableMarginBefore;
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderTableSection.cpp]: Not sure how this is related to the last baseline?

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 6]
```diff
@@ -2384,4 +2400,9 @@ void RenderFlexibleBox::layoutUsingFlexFormattingContext()
     updateLogicalHeight();
 }
 
+bool RenderFlexibleBox::childIsParticipatingInBaselineAlignment(const RenderBox& child) const
+{
+    return (alignmentForChild(child) == ItemPosition::Baseline || alignmentForChild(child) == ItemPosition::LastBaseline) && !hasAutoMarginsInCrossAxis(child) && mainAxisIsChildInlineAxis(child);
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: You could store `alignmentForChild(child)` to avoid computing it twice.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 7]
```diff
@@ -2220,30 +2225,41 @@ void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
                 continue;
             
             ItemPosition position = alignmentForChild(flexItem.box);
+            LayoutUnit maxAscent = position == ItemPosition::LastBaseline ? lineContext.lastBaselineMaxAscent : lineContext.maxAscent;
             if (position == ItemPosition::Stretch)
                 applyStretchAlignmentToChild(flexItem.box, lineCrossAxisExtent);
             LayoutUnit availableSpace = availableAlignmentSpaceForChild(lineCrossAxisExtent, flexItem.box);
-            LayoutUnit offset = alignmentOffset(availableSpace, position, marginBoxAscentForChild(flexItem.box), maxAscent, style().flexWrap() == FlexWrap::Reverse);
+            LayoutUnit offset = alignmentOffset(availableSpace, position, marginBoxAscentForChild(flexItem.box), maxAscent, style().flexWrap() == FlexWrap::Reverse, childIsParticipatingInBaselineAlignment(flexItem.box));
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: You could store `childIsParticipatingInBaselineAlignment(flexItem.box)` to avoid computing it multiple times.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 8]
```diff
@@ -2220,30 +2225,41 @@ void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
...
         minMarginAfterBaselines.append(minMarginAfterBaseline);
     }
-    
-    if (style().flexWrap() != FlexWrap::Reverse)
+
+    if (!needToAdjustItemsTowardsCrossAxis)
         return;
-    
+
+    auto adjustWrapReverseItem = [this, containerHasWrapReverse](RenderBox& box) {
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: This doesn't seem to be adjusting anything, so rename to `needToAdjustWrapReverseItem`, `shouldAdjustWrapReverseItem`, or similar.
And `box` can be `const`.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 9]
```diff
@@ -2220,30 +2225,41 @@ void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
...
-    if (style().flexWrap() != FlexWrap::Reverse)
+
+    if (!needToAdjustItemsTowardsCrossAxis)
         return;
-    
+
+    auto adjustWrapReverseItem = [this, containerHasWrapReverse](RenderBox& box) { 
+        return containerHasWrapReverse && alignmentForChild(box) == ItemPosition::Baseline && !hasAutoMarginsInCrossAxis(box); 
+    };
+    auto adjustLastBaselineItem = [this, containerHasWrapReverse](RenderBox& box) {
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: Ditto about renaming.
But why use two different functions if they will be called together with an `||`?
```cpp
auto shouldAdjustItem = [this, containerHasWrapReverse](const RenderBox& box) {
  if (containerHasWrapReverse)
    return alignmentForChild(box) == ItemPosition::Baseline && !hasAutoMarginsInCrossAxis(box);
  return alignmentForChild(box) == ItemPosition::LastBaseline && childIsParticipatingInBaselineAlignment(box);
}
```

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 10]
```diff
@@ -2220,30 +2225,41 @@ void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
...
+
+    if (!needToAdjustItemsTowardsCrossAxis)
         return;
-    
+
+    auto adjustWrapReverseItem = [this, containerHasWrapReverse](RenderBox& box) { 
+        return containerHasWrapReverse && alignmentForChild(box) == ItemPosition::Baseline && !hasAutoMarginsInCrossAxis(box); 
+    };
+    auto adjustLastBaselineItem = [this, containerHasWrapReverse](RenderBox& box) { 
+        return childIsParticipatingInBaselineAlignment(box) && alignmentForChild(box) == ItemPosition::LastBaseline && !containerHasWrapReverse;
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: Why does `adjustLastBaselineItem` check `mainAxisIsChildInlineAxis(child)` while `adjustWrapReverseItem` which is for the 1st baseline doesn't check that?

**Review by Loirooriol — COMMENTED** [review]
It sounds reasonable, but not an expert in baselines nor css-align in general.

File: Source/WebCore/rendering/RenderBlock.cpp (hunk 3 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderBlock.cpp]: Yes you are right. This should be child->previousInFlowSiblingBox()

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 10 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: This was an attempt to deal with orthogonal items. This was for:

https://wpt.fyi/results/css/css-flexbox/flexbox-align-self-baseline-horiz-006.xh%5B…%5Droduct=safari&aligned&view=interop&q=label:interop-2021-flexbox?label=experimental&label=master&aligned

The comment in the test seemed to imply that orthogonal items should be ignored, but I learned that this was based off of an old version of the spec. Instead we should be synthesizing the baselines, but that looks like it will also require an in depth discussion. For now I'll be removing that check (since it shouldn't be there anyway) and will try to synthesize the baselines correctly in another patch.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 7 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Since we no longer treat orthogonal elements as a special case (we need to synthesize the baselines instead), this argument is actually no longer needed. I went ahead and removed it from the code here.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 11]
```diff
@@ -2385,4 +2384,10 @@ void RenderFlexibleBox::layoutUsingFlexFormattingContext()
     updateLogicalHeight();
 }
 
+bool RenderFlexibleBox::childIsParticipatingInBaselineAlignment(const RenderBox& child) const
+{
+    auto alignment = alignmentForChild(child);
+    return (alignment == ItemPosition::Baseline || alignment == ItemPosition::LastBaseline) && !hasAutoMarginsInCrossAxis(child);
+}
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: This method only has a single caller and the logic is short, so adding it seems unnecessary to me.
By removing it you can reuse the `alignmentForChild(child)`.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 12]
```diff
@@ -2202,16 +2206,15 @@ void RenderFlexibleBox::adjustAlignmentForChild(RenderBox& child, LayoutUnit del
     
 void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
 {
-    // Keep track of the space between the baseline edge and the after edge of
-    // the box for each line.
-    Vector<LayoutUnit> minMarginAfterBaselines;
-    
+    bool containerHasWrapReverse = style().flexWrap() == FlexWrap::Reverse;
+    auto shouldAdjustItemTowardsCrossAxis = [this, containerHasWrapReverse](const RenderBox& box, ItemPosition alignment) { 
+        return (containerHasWrapReverse && alignment == ItemPosition::Baseline && !hasAutoMarginsInCrossAxis(box)) || (alignment == ItemPosition::LastBaseline && !containerHasWrapReverse);
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: Why check `!hasAutoMarginsInCrossAxis(box)` for the 1st baseline but not for the last one?

Comment (Loirooriol) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: If it should be checked in both cases, then the logic could be shared with another conditional:

```cpp
    auto something = [containerHasWrapReverse](ItemPosition alignment) { 
        return (containerHasWrapReverse && alignment == ItemPosition::Baseline) || (alignment == ItemPosition::LastBaseline && !containerHasWrapReverse); 
    };
// ...
            if (something(position))
                minMarginAfterBaseline = std::min(minMarginAfterBaseline, availableAlignmentSpaceForChild(lineCrossAxisExtent, flexItem.box) - offset);
// ...
                if (!hasAutoMarginsInCrossAxis(flexItem.box) && something(alignmentForChild(flexItem.box)))
                    adjustAlignmentForChild(flexItem.box, minMarginAfterBaseline);
```

Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Yes it should be for both. I made the adjustment, but let me know if you see any other issues with this

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 11 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Agreed. It was useful before, but now after the cleanup it is much less useful. Changed.

File: Source/WebCore/rendering/RenderTableSection.cpp (hunk 5 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderTableSection.cpp]: It is not directly related to the last baseline implementation but to a test case for it. There were 2 expectation files we were rendering incorrectly (css-flexbox/flexbox-align-self-baseline-horiz-001a-expected.xhtml and flexbox-align-self-baseline-horiz-001b-expected.xhtml) 

These were using the first baseline for table elements and we were not computing the correct first baseline. This caused us to fail the tests even though we were rendering the actual test file correctly. This was the incorrect way to do it and I moved the logic to RenderTable::baselinePosition. Now it is similar to other objects (e.g. RenderBlock::baselinePosition).

File: Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h (hunk 4 above)
Comment (sammygill) [reply Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h]: Actually we didn't need this anymore so it has been removed.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 13]
```diff
@@ -2051,15 +2051,21 @@ void RenderFlexibleBox::layoutAndPlaceChildren(LayoutUnit& crossAxisOffset, Vect
...
-
-            maxAscent = std::max(maxAscent, ascent);
             maxDescent = std::max(maxDescent, descent);
+            if (alignment == ItemPosition::Baseline) {
+                maxAscent =  std::max(maxAscent, ascent);
+                // FIXME: Take scrollbar into account
+                childCrossAxisMarginBoxExtent = maxAscent + maxDescent;
+            } else {
+                lastBaselineMaxAscent = std::max(lastBaselineMaxAscent, ascent);
+                childCrossAxisMarginBoxExtent = lastBaselineMaxAscent + maxDescent;
```
Comment (brentfulgham) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp:2067]: It seems like the FIXME about scrollbar is true here, too.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 14]
```diff
@@ -2221,31 +2226,27 @@ void RenderFlexibleBox::alignChildren(const Vector<LineContext>& lineContexts)
...
-            const auto& flexItem = lineContext.flexItems[childNumber];
-            if (alignmentForChild(flexItem.box) == ItemPosition::Baseline && !hasAutoMarginsInCrossAxis(flexItem.box) && minMarginAfterBaseline)
-                adjustAlignmentForChild(flexItem.box, minMarginAfterBaseline);
+        // wrap-reverse flips the cross axis start and end. For baseline alignment,
+        // this means we need to align the after edge of baseline elements with the
+        // after edge of the flex line. We can also use the same logic for last
+        // baseline alignment since those items must be pushed towards the cross
+        // start as well.
+        if (minMarginAfterBaseline) {
+            for (size_t childNumber = 0; childNumber < lineContext.flexItems.size(); ++childNumber) {
```
Comment (brentfulgham) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: I think this could be a modern C++ for loop:

```
for (auto& flexItem : lineContext.flexItems) {
  if (shouldAdjustItemTowardsCrossAxis(flexItem.box, alignmentForChild(flexItem.box)) && !hasAutoMarginsInCrossAxis(flexItem.box))
                     adjustAlignmentForChild(flexItem.box, minMarginAfterBaseline);
}
```

File: Source/WebCore/rendering/RenderTable.cpp [hunk 15]
```diff
@@ -1508,9 +1508,10 @@ RenderTableCell* RenderTable::cellAfter(const RenderTableCell* cell) const
 
 LayoutUnit RenderTable::baselinePosition(FontBaseline baselineType, bool firstLine, LineDirectionMode direction, LinePositionMode linePositionMode) const
 {
-    return valueOrCompute(firstLineBaseline(), [&] {
-        return RenderBox::baselinePosition(baselineType, firstLine, direction, linePositionMode);
-    });
+    auto baselinePos = firstLineBaseline();
+    if (baselinePos)
```
Comment (brentfulgham) [inline Source/WebCore/rendering/RenderTable.cpp]: A bit more succinct to say:

```
if (auto baselinePos = firstLineBaseline())
    return (direction == HorizontalLine ? marginTop() : marginRight()) + baselinePos.value();
```

**Review by brentfulgham — APPROVED** [review]
LGTM with the changes Loirooriol (and I) suggested.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 11 above)
Comment (Loirooriol) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: It's no longer used, remove it.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 12 above)
Comment (Loirooriol) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Remove `!hasAutoMarginsInCrossAxis(box)` from `shouldAdjustItemTowardsCrossAxis` since the caller that needs it is already checking it?

File: Source/WebCore/rendering/RenderTable.cpp [hunk 16]
```diff
@@ -1508,9 +1508,10 @@ RenderTableCell* RenderTable::cellAfter(const RenderTableCell* cell) const
...
+        return (direction == HorizontalLine ? marginTop() : marginRight()) + firstLineBaseline().value();
```
Comment (Loirooriol) [inline Source/WebCore/rendering/RenderTable.cpp]: I get that you need this fix for some tests, but it's actually fixing the reference, not the flexbox tests, so it seems strange to do it here.
I would prefer to do this change in its own PR, and have a specific test not involving flexbox, like comparing
```html
<div style="display: inline-block; background: lime;">line 1</div>
<div style="display: inline-table; margin-top: 20px; background: aqua;">line 1<br/>line 2</div>
```
with this reference:
```html
<div style="display: inline-block; margin-top: 20px; background: lime; vertical-align: top">line 1</div>
<div style="display: inline-block; margin-top: 20px; background: aqua;">line 1<br/>line 2</div>
```

**Review by Loirooriol — APPROVED** [review]
Informal LGTM with the suggested changes

File: Source/WebCore/rendering/RenderTable.cpp (hunk 16 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderTable.cpp]: That makes sense. I refactored it out and created its own PR 

https://github.com/WebKit/WebKit/pull/5153

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 13 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp:2067]: Added it to both along with a new bugzilla issue to reference and keep track of it

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 14 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Fixed! There were also a couple of other places above in the same method where I made the same change.

File: Source/WebCore/rendering/RenderTable.cpp (hunk 15 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderTable.cpp]: Changed!

**Comment by sammygill** [comment]
- [x] Once https://github.com/WebKit/WebKit/pull/5153 lands need to remove the code that was refactored out. 2 of the TestExpectations rely on that patch in order to pass

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 12 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Removed now

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 11 above)
Comment (sammygill) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Removed

**Comment by sammygill** [comment]
Rebaselined test expectations with the new tests imported from WPT
